main/src/library/RedBlackSet.flix

```diff
@@ -0,0 +1,296 @@
+///
+/// An experimental RedBlackSet type represented internally as a red-black tree.
+///
+///
+enum RedBlackSet[t] {
+    case RedBlackSet(RedBlackTree[t])
+}
+
+namespace RedBlackSet {
+
+    ///
+    /// Returns the size of `xs`.
+    ///
+    pub def size(xs: RedBlackSet[a]): Int32 =
+        let RedBlackSet(es) = xs;
+            RedBlackTree.size(es)
+
+    ///
+    /// Returns the empty set.
+    ///
+    /// `RedBlackSet#{}` is syntactic sugar for `empty` (`RedBlackSet#{} = empty()`).
+    ///
+    /// TODO: Investigate how to activate the RedBlackSet#{} syntax
+    pub def empty(): RedBlackSet[a] = RedBlackSet(Leaf)
+
+    ///
+    /// Returns the singleton set containing `x`.
+    ///
+    /// `RedBlackSet#{x}` is syntactic sugar for `singleton` (`RedBlackSet#{x} = singleton(x)`).
+    ///
+    /// TODO: Investigate how to activate the RedBlackSet#{x} syntax
+    pub def singleton(x: a): RedBlackSet[a] = insert(x, empty())
+
+    ///
+    /// Adds `x` to `xs`.
+    ///
+    pub def insert(x: a, xs: RedBlackSet[a]): RedBlackSet[a] =
+        let RedBlackSet(es) = xs;
+            RedBlackSet(RedBlackTree.insert(x, es))
+
+    ///
+    /// Removes `x` from `xs`.
+    ///
+    pub def delete(x: a, xs: RedBlackSet[a]): RedBlackSet[a] =
+        let RedBlackSet(es) = xs;
+            RedBlackSet(RedBlackTree.delete(x, es))
+
+    ///
+    /// Returns true if and only if `xs` is the empty set, i.e. `RedBlackSet(Leaf)`.
+    ///
+    pub def isEmpty(xs: RedBlackSet[a]): Bool =
+        let RedBlackSet(s) = xs;
+            match s {
+                case Leaf => true
+                case _ => false
+            }
+
+    ///
+    /// Returns true if and only if `x` is a member of `xs`.
+    ///
+    pub def memberOf(x: a, xs: RedBlackSet[a]): Bool =
+        let RedBlackSet(es) = xs;
+            RedBlackTree.contains(x, es)
+
+    ///
+    /// Returns true if and only if every element in `xs` appears in `ys`.
+    ///
+    pub def isSubsetOf(xs: RedBlackSet[a], ys: RedBlackSet[a]): Bool = forall(x -> memberOf(x, ys), xs)
+
+    ///
+    /// Returns true if and only if every element in `xs` appears in `ys` and `xs != ys`.
+    ///
+    pub def isProperSubsetOf(xs: RedBlackSet[a], ys: RedBlackSet[a]): Bool = if (isSubsetOf(xs, ys)) size(xs) != size(ys) else false
+
+    ///
+    /// Alias for `findLeft`.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def find(f: a -> Bool, xs: RedBlackSet[a]): Option[a] = findLeft(f, xs)
+
+    ///
+    /// Optionally returns the first element of `xs` that satisfies the predicate `f` when searching from left to right.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def findLeft(f: a -> Bool, xs: RedBlackSet[a]): Option[a] =
+        let result = foldLeft((acc, x) -> if (f(x)) insert(x, acc) else acc, empty(), xs);
```

Comment:
Wait. This seems needlessly slow. A find (whether find left or right) should have time O(log n) not O(n).